lib/dynarray.c

```diff
@@ -48,11 +48,23 @@
 #include "util.h"
 #include "xmalloc.h"
 
-EXPORTED struct dynarray *dynarray_new(size_t membsize)
+EXPORTED void dynarray_init(struct dynarray *da, size_t membsize)
 {
     assert(membsize);
-    struct dynarray *da = xzmalloc(sizeof(struct dynarray));
+    memset(da, 0, sizeof(struct dynarray));
     da->membsize = membsize;
+}
+
+EXPORTED void dynarray_fini(struct dynarray *da)
+{
+    free(da->data);
+    memset(da, 0, sizeof(struct dynarray));
+}
+
+EXPORTED struct dynarray *dynarray_new(size_t membsize)
+{
+    struct dynarray *da = xzmalloc(sizeof(struct dynarray));
+    dynarray_init(da, membsize);
     return da;
 }
 
```

Comment:
How is dynarray different from the existing ptrarray?

Reply:
Does `dynarray_new()` call twice `memset(da, 0, sizeof(struct dynarray))` on the same data in a row: once in xzmalloc() and once in `dynarray_init()` (called from dynarray_new)?

Reply:
Is is more cache-friendly than ptrarray. A ptrarray requires to dereference a pointer for each array member.  A dynarray keeps a contiguous memory block of same-sized elements, so it's faster to traverse them.

Reply:
In addition, it allows us to save n malloc/free calls for an array of n members

Reply:
@dilyanpalauzov right, I now changed the `dynarray_new` function to just `xmalloc`, rather than `xzmalloc`.